include/nbl/builtin/hlsl/bxdf/cook_torrance_base.hlsl

```diff
@@ -17,95 +17,303 @@ namespace hlsl
+        quant_type DG1 = ndf.template DG1<sample_type, isotropic_interaction_type>(dq, qq, _sample, interaction);
+        return DG1.projectedLightMeasure;
     }
-    template<class Query>
-    MT __DG1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(anisocache_type) cache)
+    scalar_type pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
     {
-        MT measure_transform;
-        measure_transform.pdf = ndf.template DG1<Query>(query, cache);
-        measure_transform.transmitted = cache.isTransmission();
-        measure_transform.VdotH = cache.getVdotH();
-        measure_transform.LdotH = cache.getLdotH();
-        measure_transform.VdotHLdotH = cache.getVdotHLdotH();
-        return measure_transform;
+        using quant_query_type = typename N::quant_query_type;
+        using dg1_query_type = typename N::dg1_query_type;
+
+        scalar_type dummy;
+        quant_query_type qq = ndf.template createQuantQuery<anisocache_type>(cache, dummy);
+        dg1_query_type dq = ndf.template createDG1Query<anisotropic_interaction_type, anisocache_type>(interaction, cache);
+        quant_type DG1 = ndf.template DG1<sample_type, anisotropic_interaction_type>(dq, qq, _sample, interaction);
+        return DG1.projectedLightMeasure;
     }
 
-    template<class Query>
-    MT __DG(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    quotient_pdf_type quotient_and_pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
     {
-        MT measure_transform;
-        measure_transform.pdf = ndf.template D<isocache_type>(cache);
-        NBL_IF_CONSTEXPR(MT::Type == ndf::MicrofacetTransformTypes::MTT_REFLECT_REFRACT)
-            measure_transform.transmitted = cache.isTransmission();
-        NBL_IF_CONSTEXPR(MT::Type == ndf::MicrofacetTransformTypes::MTT_REFRACT)
+        scalar_type _pdf = pdf(_sample, interaction, cache);
+
+        spectral_type quo = hlsl::promote<spectral_type>(0.0);
+        if (_sample.getNdotL() > numeric_limits<scalar_type>::min && interaction.getNdotV() > numeric_limits<scalar_type>::min)
         {
-            measure_transform.VdotH = cache.getVdotH();
-            measure_transform.LdotH = cache.getLdotH();
-            measure_transform.VdotHLdotH = cache.getVdotHLdotH();
+            using g2g1_query_type = typename N::g2g1_query_type;
+
+            g2g1_query_type gq = ndf.template createG2G1Query<sample_type, isotropic_interaction_type>(_sample, interaction);
+            scalar_type G2_over_G1 = ndf.template G2_over_G1<sample_type, isotropic_interaction_type, isocache_type>(gq, _sample, interaction, cache);
```

Comment:
`G2_over_G1` probably needs a if-check of `ndf.isSmooth()` before calling and not do the tests inside itself for that (so we can skip constructing the queries)